doc/source/raysgd/v2/user_guide.rst

```diff
@@ -126,42 +135,121 @@ Porting code to RaySGD
 
         TODO. Write about how to convert code to use horovod.
 
+.. To make existing code from the previous SGD API, see :ref:`Backwards Compatibility <sgd-backwards-compatibility>`.
 
+Configurations
+--------------
 
-Training on a large dataset
----------------------------
+With RaySGD, you can execute a training function (``train_func``) in a
+distributed manner by calling ``trainer.run(train_func)``. To pass arguments
+into the training function, you can expose a single ``config`` parameter:
 
-SGD provides native support for :ref:`Ray Datasets <datasets>`. You can pass in a Dataset to RaySGD via ``Trainer.run``.
-Underneath the hood, RaySGD will automatically shard the given dataset.
+.. code-block:: diff
 
+    -def train_func():
+    +def train_func(config):
+
+Then, you can pass in the config dictionary as an argument to ``Trainer.run``:
+
+.. code-block:: diff
+
+    -trainer.run(train_func)
+    +config = {} # This should be populated.
+    +trainer.run(train_func, config=config)
+
+Putting this all together, you can run your training function with different
+configurations. As an example:
 
 .. code-block:: python
 
+    from ray.util.sgd.v2 import Trainer
+
     def train_func(config):
-        batch_size = config["worker_batch_size"]
-        data_shard = ray.sgd.get_data_shard()
-        dataloader = data_shard.to_torch(batch_size=batch_size)
+        results = []
+        for i in range(config["num_epochs"]):
+            results.append(i)
+        return results
 
-        for x, y in dataloader:
-            output = model(x)
-            ...
+    trainer = Trainer(backend="torch", num_workers=2)
+    trainer.start()
+    print(trainer.run(train_func, config={"num_epochs": 2}))
+    # [[0, 1], [0, 1]]
+    print(trainer.run(train_func, config={"num_epochs": 5}))
+    # [[0, 1, 2, 3, 4], [0, 1, 2, 3, 4]]
+    trainer.shutdown()
 
-        return model
+A primary use-case for ``config`` is to try different hyperparameters. To
+perform hyperparameter tuning with RaySGD, please refer to the
+:ref:`Ray Tune integration <tune-sgd>`.
 
-    trainer = Trainer(num_workers=8, backend="torch")
-    dataset = ray.data.read_csv("...").filter().pipeline(length=50)
+.. TODO add support for with_parameters
 
-    result = trainer.run(
-        train_func,
-        config={"worker_batch_size": 64},
-        dataset=dataset)
+.. _sgd-logging:
 
+Logging
```

Comment:
Hmm, this section only refers to logging though. Should I convert these all into subsections within a `Logging, Monitoring, and Callbacks` section or is that too granular?

Reply:
Yeah I'm thinking about this some more and to me `sgd.report()`, Callbacks, Logging Callbacks, and the `run_iterator` API are all related and I think makes sense to have as subsections of a single section.

As a user it might get confusing to have separate sections on these things, when really I have to understand these concepts together to use them end-to-end. Reading just the callbacks section is not going to be useful if I don't learn about `sgd.report()` and vice versa.

What about something like this:
Directory Structure
Logging, Monitoring, and Callbacks
    1. Reporting Intermediate Results
    2. Callbacks
    3. Custom Logging & Monitoring
        3a. Custom callbacks
        3b. Link to run_iterator API and show mlflow example

Reply:
Yeah that makes sense! I kept `Custom Callbacks` directly under `Callbacks` since I think it's more natural to read, and didn't add the `Custom Logging & Monitoring` section for the `run_iterator` example, maybe because right now I'm not yet convinced we want to highlight this API... even the `mlflow` example looks like it can be done in a callback.

Reply:
Yeah it can be done in a callback, but the callback doesn’t exist yet. I think it’s important to highlight this to show users that it’s relatively easy to do this themselves and they don’t need to wait for us to implement certain callbacks in order to have this functionality.

Reply:
If I have a workflow with distributed Pytorch and MLflow, I shouldn’t not use Ray SGD just because there’s no mlflow callback. Because it’s not too difficult to do this with the run_iterator API.

Same thing can generalize to library X.

Reply:
I just think this isn’t something we should try to hide and can actually be useful for users.

Reply:
Added a TODO section for `Advanced Customization` based on the iterator API. Will add this along with the section for the iterator API in a future PR, but in general this should explain when to use this API over defining a custom callback (e.g. for more complex integrations with other libraries).